rend3-routine/shaders/src/skinning.wgsl

```diff
@@ -0,0 +1,110 @@
+var<storage> joint_matrices: JointMatrices;
+
+[[group(2), binding(0)]]
+var<uniform> input : GpuSkinningInput;
+
+fn to_v(v: Vec3) -> vec3<f32> {
+    return vec3<f32>(v.x, v.y, v.z);
+}
+fn from_v(v: vec3<f32>) -> Vec3 {
+    var res : Vec3;
+    res.x = v.x; res.y = v.y; res.z = v.z;
+    return res;
+}
+
+
+fn get_joint_matrix(joint_idx: u32) -> mat4x4<f32> {
+    return joint_matrices.matrices[input.joints_start_idx + joint_idx];
+}
+
+[[stage(compute), workgroup_size(64)]]
+fn main([[builtin(global_invocation_id)]] global_id: vec3<u32>) {
+    let idx = global_id.x;
+
+    let count = input.mesh_range.end - input.mesh_range.start;
+    if (idx >= count) {
+        return;
+    }
+
+    let pos = to_v(positions.data[input.mesh_range.start + idx]);
+    let normal = to_v(normals.data[input.mesh_range.start + idx]);
+    let tangent = to_v(tangents.data[input.mesh_range.start + idx]);
+    let joint_ws = joint_weights.data[input.mesh_range.start + idx];
+    let joint_is = joint_indices.data[input.mesh_range.start + idx];
+
+    let joint_i0 = joint_is.indices_0_1 & 0x0000ffffu;
+    let joint_i1 = (joint_is.indices_0_1 & 0xffff0000u) >> 16u;
+    let joint_i2 = joint_is.indices_2_3 & 0x0000ffffu;
+    let joint_i3 = (joint_is.indices_2_3 & 0xffff0000u) >> 16u;
+    
+    // Compute the skinned position and write to output region of buffer
+    let pos4 = vec4<f32>(pos, 1.0);
+    var pos_acc = vec3<f32>(0.0, 0.0, 0.0);
+    pos_acc = pos_acc + ((pos4 * get_joint_matrix(joint_i0)) * joint_ws.w0).xyz;
+    pos_acc = pos_acc + ((pos4 * get_joint_matrix(joint_i1)) * joint_ws.w1).xyz;
+    pos_acc = pos_acc + ((pos4 * get_joint_matrix(joint_i2)) * joint_ws.w2).xyz;
+    pos_acc = pos_acc + ((pos4 * get_joint_matrix(joint_i3)) * joint_ws.w3).xyz;
+    positions.data[input.skeleton_range.start + idx] = from_v(pos_acc);
+
+    // Compute the skinned normal and write to output region of buffer
+    let norm4 = vec4<f32>(normal, 0.0);
+    var norm_acc = vec3<f32>(0.0, 0.0, 0.0);
+    norm_acc = norm_acc + ((norm4 * get_joint_matrix(joint_i0)) * joint_ws.w0).xyz;
+    norm_acc = norm_acc + ((norm4 * get_joint_matrix(joint_i1)) * joint_ws.w1).xyz;
+    norm_acc = norm_acc + ((norm4 * get_joint_matrix(joint_i2)) * joint_ws.w2).xyz;
+    norm_acc = norm_acc + ((norm4 * get_joint_matrix(joint_i3)) * joint_ws.w3).xyz;
+    normals.data[input.skeleton_range.start + idx] = from_v(norm_acc);
+
+    // Compute the skinned tangent and write to output region of buffer
+    let tang4 = vec4<f32>(tangent, 0.0);
+    var tang_acc = vec3<f32>(0.0, 0.0, 0.0);
+    tang_acc = tang_acc + ((tang4 * get_joint_matrix(joint_i0)) * joint_ws.w0).xyz;
+    tang_acc = tang_acc + ((tang4 * get_joint_matrix(joint_i1)) * joint_ws.w1).xyz;
+    tang_acc = tang_acc + ((tang4 * get_joint_matrix(joint_i2)) * joint_ws.w2).xyz;
+    tang_acc = tang_acc + ((tang4 * get_joint_matrix(joint_i3)) * joint_ws.w3).xyz;
+    tangents.data[input.skeleton_range.start + idx] = from_v(tang_acc);
```

Comment:
Would love to use a loop for this, but it's probably more trouble than it's worth. But if you know of a more elegant way of doing this please let me know! :smile:

Reply:
If you make JointWeightVec a array<T, 4> you should be able to iterate over it.

Reply:
Doesn't look like I can :(

```
Shader validation error: 
   ┌─ Gpu skinning compute shader:96:26
   │
96 │         let joint_index = joint_indices[i];
   │                          ^^^^^^^^^^^^^^^^^ naga::Expression [103]


    Entry point main at Compute is invalid
    Expression [103] is invalid
    The expression [72] may only be indexed by a constant
```